pm: policy: latency: Add support for immediate action

nordic-krch opened the pull request:
Remove API for subscription to the latency change. This API was added to support executing immediate action on latency change but implementation was calling subscriber callbacks with interrupts locked which is not desirable. In general, it is non-trivial task to manage multiple requests and perform asynchronous action for the top request. For example, on-off manager implements that for binary case.

Additionally, API was missing a notification to the requester when immediate action is completed and it is required for the user to know when requested latency requirement is fulfilled.

Proposed API is adding a possible notification for request and update and adds option to add a manager for controlling immediate action. Added notifications make sense only if immediate action is used. Immediate action implements support for binary case where there is a single action that is performed when latency is below a given threshold. In future it can be extended to support gradual manager.

Latency policy API was added #42040 #48476 but never used beyond test and sample. One of the use cases is MRAM latency management (#81381) and current API could not handle that case for reasons described above.

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +267,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
  * The system will not enter any power state that would make the system to
  * exceed the given latency value.
  *
+ * If immediate control manager is provided request may trigger an action asynchronous action
```
What's the "immediate control manager"? I didn't find the term in Zephyr docs.

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +267,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
  */
-void pm_policy_latency_request_add(struct pm_policy_latency_request *req,
-				   uint32_t value_us);
+int pm_policy_latency_request_add_sync(struct pm_policy_latency_request *req, uint32_t value_us);
```
I would say the default function should be synchronous (without a suffix) and the special one, callable from IRQs, should be `_async()`.

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +267,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
  * @param req Latency request.
  * @param value_us Maximum allowed latency in microseconds.
+ *
+ * @retval 0 if request is applied.
+ * @retval 1 if request required immediate action that is not completed. Configured completion
+ * notification will inform about completion.
```
In async functions, I would expect that the callback is called regardless of the returned value. So this could be simplified to `retval 0 if the request is pending. When the callback is applied, the callback notification is called with the result`.

nordic-krch committed:
tests: pm: policy_api: Add test for latency immediate action

Add test for the case where latency change triggers an immediate
action. Configuration is using a single threshold for triggering
of the action.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +267,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
+ */
+int pm_policy_latency_request_remove(struct pm_policy_latency_request *req);
+
+/** @brief Immediate action manager for single threshold.
```
The "immediate" word used around max latency changes subscription is confusing to me. Is there any "delayed max latency changes subscription" to differentiate the "immediate" one from?

Maybe it would be enough to name this structure `pm_policy_latency_change_binary_subscription`?

hubertmis replied:
I don't fully agree. When latency is used to select a CPU sleep state, the immediate action is locking a set of CPU sleep states that are characterized by greater latency. I agree, that the visible effect of this action is only when the CPU enters the idle thread, but we can treat locking as an immediate action.

The same way with NVM controllers: the immediate action is locking their sleep states that are characterized by greater latency. But this action has visible effect after the NVM inactivity timeout fires.

So, you can treat both of these actions as "immediate" - locking states are immediate in both, or as "deferred" - the actual sleep states are entered after inactivity is detected.

The main difference between these two is that CPU-idle handling is CPU-specific, while NVM-idle handling is SoC-specific or even PCB-specific if an NVM controller is external.

nordic-krch committed:
pm: policy: latency: Add support for immediate action

Remove API for subscription to the latency change. This API was added
to support executing immediate action on latency change but
implementation was calling subscriber callbacks with interrupts locked
which is not desirable. In general, it is non-trivial task to
manage multiple requests and perform asynchronous action for the
top request. On-off manager implements that for binary case.

Additionally, API was missing a notifiction to the requestor when
immediate action is completed and it is required for the used to
know when requested latency requirement is fulfilled.

Proposed API is adding a possible notification for request and
update and adds option to add a manager for controlling immediate
action. Added notifications make sense only if immediate action
is used. Immediate action implements support for binary case
where there is a single action that is performed when latency
is below a given threshold. In future it can be extended to
support gradual manager.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -27,37 +28,29 @@ extern "C" {
 	sys_snode_t node;
-	pm_policy_latency_changed_cb_t cb;
+	uint32_t value_us;
+	struct onoff_client cli;
+	void *internal;
+	uint32_t flags;
```
`onoff_client` and two 32-bit variables per latency requestor seem like high memory overhead. Wouldn't it be enough to track these data per subscriber instead of per requestor?

nordic-krch replied:
if immediate action/subscription is needed then it is required. We can add kconfig option which enables that feature and have those fields conditionally compiled.

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +268,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
+	int32_t thr;
+};
+
+struct pm_policy_latency_immediate_ctrl {
```
And this one `pm_policy_latency_change_subscription`?

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -275,63 +268,140 @@ static inline int32_t pm_policy_next_event_ticks(void)
  */
-void pm_policy_latency_changed_subscribe(struct pm_policy_latency_subscription *req,
-					 pm_policy_latency_changed_cb_t cb);
+int pm_policy_latency_immediate_ctrl_add(struct pm_policy_latency_immediate_ctrl *mgr);
```
And restore the name of this function? The previous one was much clearer to me.

hubertmis reviewed (commented):
I reviewed only the policy.h API. I think we should decide on changes there first.
